src/charm.py

```diff
@@ -0,0 +1,323 @@
+                  cmd,
+                  return_code,
+                  result.stdout,
+                  result.stderr,
+         )
+
+    def _run_migrate_synapse(self):
+        """Runs the migrate command for synapse"""
+        Result = collections.namedtuple(
+                "CommandExecResult", "return_code stdout stderr")
+        cmd = ["/start.py", "migrate_config"]
+        process = self._container().exec(
+            cmd,
+            working_dir="/data",
+            environment=self._populate_synapse_env_settings(),
+        )
+        try:
+            stdout, stderr = process.wait_output()
+            result = Result(
+                    return_code=0, stdout=stdout, stderr=stderr)
+        except ops.pebble.ExecError as error:
+            result = Result(error.exit_code, error.stdout, error.stderr)
+        return_code = result.return_code
+        logger.debug(
+                  "Run command: %s, return code %s\nstdout: %s\nstderr:%s",
+                  cmd,
+                  return_code,
+                  result.stdout,
+                  result.stderr,
+         )
+
+    def _current_synapse_config(self):
+        """Retrieve the current version of /data/homeserver.yaml from server.
+
+        return None if not exists.
+
+        Returns:
+        The content of the current homeserver.yaml file, str.
+        """
+        synapse_config_path = self._SYNAPSE_CONFIG_PATH
+        container = self._container()
+        if container.exists(synapse_config_path):
+            return self._container().pull(synapse_config_path).read()
+        return None
+
+    def _on_config_changed(self, event):
+        """Handle changed configuration.
+
+        Change this example to suit your needs.
+        If you don't need to handle config, you can remove
+        this method.
+
+        Learn more about config at https://juju.is/docs/sdk/config
+        """
+        # Fetch the new config value
+        log_level = self.model.config["log-level"].lower()
+
+        # Do some validation of the configuration option
+        if self._stored.db_password is not None:
+            if self._current_synapse_config() is None:
+                self._run_migrate_synapse()
+            # The config is good, so update the configuration of the workload
+            container = self.unit.get_container(self._CONTAINER_NAME)
+            # Verify that we can connect to the Pebble API
+            # in the workload container
+            if container.can_connect():
+                # Push an updated layer with the new config
+                container.add_layer(
+                        self._CONTAINER_NAME, self._pebble_layer, combine=True)
+                container.replan()
+
+                logger.debug(
+                        "Log level for synapse changed to '%s'", log_level
+                        )
+                self.unit.status = ActiveStatus()
+            else:
```

Comment:
Looks like a bit of nesting here, should we reduce that, e.g., using a sub function?

Reply:
I'd also start with if "not can_connect" to exit the non happy cases first same for if "db_password is None"